### DISCUSSION TITLE
Allow user to add notes to habits

### DISCUSSION THREAD

**Discussion by iSoron** [iSoron/uhabits#56] [Feature Requests] [answered]
Currently, habits have titles and questions for the reminders. It would be nice to have a general "notes" field to allow the user to store whatever other information they would like about the habit, such as the reasons. These notes would be optional, and displayed only on the top of the statistics page.

This covers issues #54 and #47.

**Comment by LuccoJ** [comment]
I think it would be also nice to additionally store an annotation for each specific occurrence of the habit (with the default, of course, being no annotation).

This way you could say: "Yes, I've jogged!" "... but for some minutes less than my target is, because I was interrupted", or, for instance, I'm thinking of the way glucose meters that diabetics use work, which allow you to mark each reading with a "pre-prandial", "post-prandial" or "see diary for notes" icon.

As to the UI side of this, I use short taps for ticking, so in my use case, it would be natural for long tap to cause the annotation requester to show up. Not sure how to deal with it when long tap is the default.

**Comment by jwleon** [comment]
I second that idea for the following reason: Sometimes, there are things that you are quit "proud of" after finishing. This could be "first 5 miles run", "cracked the 150lbs mark" or, in my case, "finished book XYZ". It would be a great psychological benefit to not only see that you continued your struggle, but that you had a specific success at a specific day.

I hate seeing those things as a set-in-stone pre-defined mile-stone that you either reach or fail to reach and would rather keep this flexibel like an "annotation" to your usual habit-development.

**Comment by GordianDziwis** [comment]
Why not just make the questions field a more general notes field and add markdown support?
This would be less opiniated and would not add any new ui-element.

**Comment by Lvl4Sword** [comment]
I don't see this as hard to implement, and would make it a program I could recommend to others.
As-is this is a limiting factor to a few people.

**Comment by ralfkr** [comment]
@iSoron I am interested in implementing this feature as you have described it (i.e.: an optional "notes" field) as a way for me to familiarize myself with the app's code.  Feel free to assign it to me and I'll work on it this week.

**Comment by padluq** [comment]
I wholeheartedly support LuccoJ's request.  In fact, at this point the I feel this is the only advantage that Habit Bull has over Loop Habits.  Thanks!

-Adrian

> I think it would be also nice to additionally store an annotation for each specific occurrence of the habit (with the default, of course, being no annotation).
> 
> This way you could say: "Yes, I've jogged!" "... but for some minutes less than my target is, because I was interrupted", or, for instance, I'm thinking of the way glucose meters that diabetics use work, which allow you to mark each reading with a "pre-prandial", "post-prandial" or "see diary for notes" icon.
> 
> As to the UI side of this, I use short taps for ticking, so in my use case, it would be natural for long tap to cause the annotation requester to show up. Not sure how to deal with it when long tap is the default.

**Comment by recheej** [comment]
@iSoron. Thinking this one through....

Honestly I think this could be marked as closed. Right now the "question" field is only known as question as the hint in the xml, and the default values you present are in question form. Other than that... it's referred to as "description" in the rest of the code, including the sql column. 

I think you could make it even more generic by simply:

1. Make the edittext much bigger. Like maybe it fills the rest of the screen?
2. Possibly make the SQL column size/length bigger. Haven't taken a look... but maybe the default is pretty small right now?
3. Update your hint to make it obvious to the user that this is field can be used for anything.

And that's it. Like I said, it is pretty much already a notes/description.

**Comment by iSoron** [comment]
@recheej I think we should create a new "question" field in the database, move the current description there, then add some interface elements to allow the user to see/edit the actual description. Having one field to store two different pieces of information (question+notes) is a bit confusing, and may restrict our options in the future (for example, if we want to present the question in some other way).

**Comment by iSoron** [comment, marked as answer]
This feature has been implemented and will become available in Loop 2.0.

**Reply by Mrnofish** [reply]
Is there a way to attach a note to a single day, or just a general additional description field that belongs to the habit as a whole? I'm guessing the latter but wanted to make sure.

**Reply by iSoron** [reply]
Currently, you can only add a note to the habit as a whole. Feel free to upvote #194.

**Reply by Mrnofish** [reply]
Done, looks like nobody else has upvoted yet; nonetheless there are 22 positive reactions to the OP, perhaps because at the time upvoting wasn't possible.

**Comment by mstoetzer** [comment]
May I add something to this issue? I'm super unsure, how much effort it will be to support this notes field to be in Markdown. (like habitica is allowing it in notes in tasks). The editor itself does not do any Markdown Syntax Highlighting during editing mode.

It could be difficult to render e.g. bullet lists without breaking the UI or design of Loop. I'm just typing my thoughts here. What do you think?

**Reply by Mrnofish** [reply]
They're collecting upvotes in the Discussions section: https://github.com/iSoron/uhabits/discussions/194

Personally I don't think highlighting or wysiwyg are necessary for short daily notes concerning a habit, but that's just my opinion.

Particularly because markdown is designed to maintain readability even when used in the plainest of text editors.

I guess if someone is interested in implementing any of that, they can probably make it happen, as long as the maintainer also agrees.
